### PR TITLE
Feat/migration ipfs download

### PR DESCRIPTION
**Pull request by gammazero**
go-ipfs can now fetch migrations via IPFS

Project proposal: https://github.com/protocol/web3-dev-team/pull/86

### CODE REVIEW

File: repo/fsrepo/migrations/ipfsfetcher_test.go
Comment (mvdan) [inline repo/fsrepo/migrations/ipfsfetcher_test.go:10]: maybe hide it behind a flag, then :) otherwise the only way to run it is by editing the source.

is there any way to eventually make this test work in a local way without extra setup? e.g. by setting up some IPFS/HTTP servers/nodes.

File: repo/fsrepo/migrations/ipfsfetcher_test.go
Comment (mvdan) [inline repo/fsrepo/migrations/ipfsfetcher_test.go:24]: I think this would be clearer if called `lines` :)

**Review by mvdan — COMMENTED** [review]
Just did a first pass as it's quite late. I can do a second one tomorrow.

File: repo/fsrepo/migrations/ipfsfetcher_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher_test.go]: done

File: repo/fsrepo/migrations/ipfsfetcher_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher_test.go]: Done - hidden behind flag

For the HttpFetcher I created a dummy test server.  I am not really sure there is a great way to do this for IPFS.  Mocking interfaces seems a bit excessive.

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: Done, all errors returned

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: Right, that is a mistake.

Done - use path instead of filepath

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: Will print to stderr for now, and make it look less like an error.  If ipfsfetcher is only going to be used from within go-ipfs, then it may be better to not try to load the pluging at all.

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: done

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: done

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: distPath is a URL path, so I will treat it as such.  The use of filepath on it was a mistake.

File: repo/fsrepo/migrations/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher.go]: done

**Comment by gammazero** [comment]
@mvdan I addressed all review comments and made suggested changes.  One other big change I needed to make was to move IpfsFetcher into its own package.  This is so that things outside of ipfs, that import `migrations`, are not forced to link with all the IpfsFetcher baggage.

**Comment by gammazero** [comment]
Remaining items:
- Migration policy config
- Node should not listen for inbound connections
- Node should not expose an API/gateway.
- HTTP fetcher => HTTPS fetcher
- Read IPFS peers from existing config

File: repo/fsrepo/migrations/ipfsfetcher_test.go
Comment (mvdan) [reply repo/fsrepo/migrations/ipfsfetcher_test.go]: I'm just a bit worried that we have a test that will practically never be run. Mocking isn't great, but if running a real IPFS node for the test is too difficult, at least testing with a mock by default would be better than testing nothing by default. The flag could still be used to test against a real IPFS node instead of a mock.

Comment (iand) [reply repo/fsrepo/migrations/ipfsfetcher_test.go]: Another pattern is to use an environment variable and put the name of it in the Skip message. Inspired by this blog post https://peter.bourgon.org/blog/2021/04/02/dont-use-build-tags-for-integration-tests.html

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:160]: We could do that, but then if you try and import the data into the new node from the old node for seeding purposes you're going to need to either:
- Use the import/export APIs (e.g. use CAR files to move the data around)
- Have the temporary node connect to the new node so it can pull the data using Bitswap

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:232]: Why would we get an HTTP(S) URL here? If we can't then let's remove this code, if we can then can let's discuss our options since this seems it will get us into trouble (e.g. what about subdomain gateways, maybe path gateways can start with some prefix like foo.tld/custom/ipfs/Qmxyz).

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go:14]:
```suggestion
		t.Skip("manually-run dev test, use the '-ipfstest' flag to run")
```

Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go:14]: Is the reason this test is manual because it makes network requests?

**Review by aschmahmann — CHANGES_REQUESTED** [review]

File: cmd/ipfs/daemon.go
Comment (aschmahmann) [reply cmd/ipfs/daemon.go:52]: The proposal https://github.com/protocol/web3-dev-team/pull/86/files#diff-991744cab1b93840a21929cbf5aa7bc00a37ae4244b5c569df086214e36e47cfR98 indicates we should just do HTTP first and fallback to IPFS for a release or two and then switch to IPFS first.

Are we running into testing issues doing it that way?

Comment (gammazero) [reply cmd/ipfs/daemon.go:52]: This was done as a way to test/demo before and [configurable policy](https://github.com/protocol/web3-dev-team/pull/86/files#diff-991744cab1b93840a21929cbf5aa7bc00a37ae4244b5c569df086214e36e47cfR66) is available.  I have an update that will provide this policy and will remove the migrate `ipfs` flag.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:160]: Since the files are already available on local disk (since they need to be run), it is just as easy for me to add the files to the real node after migration. So, I think this is fine since I am not using the temp node for anything other than downloading migrations.

Comment (aschmahmann) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:160]: > it is just as easy for me to add the files to the real node after migration

Right, but if we do this it needs to be as a CAR file not a UnixFS import of some binary since we want the CIDs to match and `ipfs add <file>` is not guaranteed to be the same across versions and boxing ourselves into some parameters now wouldn't be a good idea either. i.e. go-ipfs needs to understand "I am importing this DAG" not "I am importing this binary", whether it's a CAR file via DAG import, or UnixFS via Bitswap is 🤷.

File: cmd/ipfs/daemon.go
Comment (gammazero) [reply cmd/ipfs/daemon.go]: It seems like allowing an entire config to be specified for a temp node also invites many problems, as users will not necessarily know what the new configuration looks like until after the migration and may typically just use their existing config (leading to issues mentioned above).  Also, the configuration for a temp, used only to download a migration or two, does not need more than a default config.  Users not knowing this may think it best to use a config similar to what they are using for their real node.

Instead of a config file, WDYT about letting the user specify peer(s) on the command line, if they happen to know one 
 or more that may be useful for migration?

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go]: Returns list of errors if all fetchers fail.  Switched to using `multierror`

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go]: Removed.  If someone were to try to use the IpfsFetcher with an HTTP path, then returning an is acceptable, and they really mean to use the HttpFetcher.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go]: Fixed

Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go]: Yes, and I did not know a good way to simulate or mock an IPFS node, but I wanted to have some test that a developer could run to check if things are working.  WDYT? Leave it, toss it, create a better test?

File: cmd/ipfs/daemon.go
Comment (aschmahmann) [reply cmd/ipfs/daemon.go]: > It seems like allowing an entire config to be specified for a temp node also invites many problems...

Ya, it's using a pretty big hammer to deal with a smaller set of problems but needing to modify the config for the purposes of migrations at all seems like a potentially advanced requirement for an end user, perhaps sufficiently so that they might as well get access to the full range of tools.

However, there are also potentially a large number of options that could be specified during a migration so idk if command line flags for each of them is really a reasonable alternative. For example, if someone is on a private network and therefore needs the network symmetric key they're a bit out of luck with just some peer flags. If we make our routing systems more configurable we're similarly out of luck.

cc @Stebalien as this and https://github.com/ipfs/go-ipfs/pull/8064#discussion_r611867631 seem to be related to the comments in the proposal PR (https://github.com/protocol/web3-dev-team/pull/86/files#diff-991744cab1b93840a21929cbf5aa7bc00a37ae4244b5c569df086214e36e47cfR66 and https://github.com/protocol/web3-dev-team/pull/86/files#r610934803)

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go
Comment (aschmahmann) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go]: There are some tests in the integration test package (e.g. https://github.com/ipfs/go-ipfs/blob/master/test/integration/three_legged_cat_test.go) that could be of inspiration here.

If it's a pain to write the test lmk and we can decide on an alternative. Using a flag/env var seems fine (I'm not sure if we have an appropriate one, although perhaps this one https://github.com/ipfs/go-ipfs/blob/ef866a1400b3b2861e5e8b6cc9edc8633b890a0a/test/integration/addcat_test.go#L173 would do), although I don't really trust tests that don't run in CI since they aren't reliably run before merging PRs and are subject to breaking.

@Stebalien lmk if you have any additional thoughts here given you've been thinking about our CI situation a bit, or it's just 👍.

Comment (Stebalien) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go]: We need to test this in CI but we can use sharness tests. If a test is disabled in CI, it might as well not exist.

File: repo/fsrepo/migrations/fetcher.go
Comment (Stebalien) [reply repo/fsrepo/migrations/fetcher.go]: FYI, https://github.com/uber-go/multierr/ nicer in my experience (doesn't matter too much).

File: cmd/ipfs/daemon.go
Comment (Stebalien) [reply cmd/ipfs/daemon.go]: > An alternative way to do this is to have a --migration-temp-node-config (or better name blush) option that takes a config file to use for the temporary node. This could protect us from a variety of bugs/corner cases as it allows users to configure the temporary node as if it was a regular one.

We _could_, maybe, later, support that for advanced users/distros. That's not going to really help in this case because users will be _very_ confused about why their migration options aren't being applied. Users are going to be less confused if we start a temporary node that ignores their config.

---

I've left a comment on the proposal. IMO, we should just try to read the relevant part of the config and ignore the rest. Downsides:

1. Migrations can't touch the migration part of the config. Honestly, I'm fine with that.
2. If we decide to move the config, things could get... tricky. But trying a set of configs in order of precedence, _just_ to read a small portion probably isn't a huge issue.

We could also do this on the command-line, but it could get icky (as @aschmahmann said, lots of options).

File: cmd/ipfs/migration.go
Comment (aschmahmann) [inline cmd/ipfs/migration.go:108]: This is the thing that doesn't work https://github.com/ipfs/go-ipfs/pull/8064#discussion_r611921297.

For example, if we add the migration binaries using the current defaults for `ipfs add` and then change the defaults in a subsequent release then the swarms won't converge.

Comment (aschmahmann) [reply cmd/ipfs/migration.go:108]: Note: @gammazero if this is a pain to implement for you given time constraints I'm happy to either drop this requirement for the time being or I'll just code up the CAR import/export stuff on my own (likely requires copy-pasting code from the commands package into a function on the IpfsNode and then just calling it).

File: cmd/ipfs/daemon.go
Comment (aschmahmann) [reply cmd/ipfs/daemon.go]: I think I'm fine with this approach. If we define the migrations config options as described in the proposal and then we try to deserialize the config file (or object) into a struct that only has the migration values that should be fine.

This will mean that unlike the rest of the config which is nominatively typed in that backwards incompatible changes are tied to a repo version, that we are likely to use structural typing for the migrations specific section (although if this proves too hard we can add a field into the migrations struct to track versioning).

This reliance on structural typing means we basically don't have to worry too much about the potential migration config changes for the time being and it's possible we'll end up with a repo/package that contains all the migration configs (like what we used to do with fs-repo-migrations, but really small since it's only for a portion of the config file).

For the time being we can probably punt on dealing with private swarms or people with custom bootstrap nodes since they can:
1. Just use HTTP, like they've been doing so far
2. Download the migrations binary manually, or run the full fs-repo-migrations binary or ipfs-update

If this becomes a problem and we need the sledge hammer of passing in `--migration-temp-node-config` we can do that, but let's delay until we have a user + use case who needs this.

File: cmd/ipfs/daemon.go
Comment (gammazero) [reply cmd/ipfs/daemon.go]: removed

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:192]: - When migrations are downloaded via IPFS, add the migration archives to IPFS by connecting to the temporary migration node and getting the migrations over IPFS.
- When migrations are downloaded via HTTP, add the migration archive files directly to IPFS.  After talking to @Stebalien he thought it was still worth to add the files directly, even though it might result in a different dag.  If we decide later to not do that, it is a trivial change.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go]: @aschmahmann I will see what I can do with sharness, but likely this is going to have to become a job for your team if it needs to get done before the upcoming release.

File: cmd/ipfs/migration.go
Comment (gammazero) [reply cmd/ipfs/migration.go]: I went with the approach of connecting to the temporary migration node and then getting the data over IPFS.  Exposing a CAR interface on the node may be a better solution in the long run, and we can revisit that.

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go]: daemon was already using https://github.com/hashicorp/go-multierror. 🤷

File: cmd/ipfs/daemon.go
Comment (gammazero) [reply cmd/ipfs/daemon.go]: I am going to leave the configuration for a later time.  Currently, go-ipfs accepts a couple of optional flags that allow the user to specify how to download the migrations (`--download-migration`) and if/how to keep downloaded migrations (`--keep-migration`).  Even if this information is available from a config file in the future, I think these flags will will still be useful to override that.

**Comment by gammazero** [comment]
Rebased on master to pick up fixed sharness tests.

File: cmd/ipfs/daemon.go
Comment (Stebalien) [reply cmd/ipfs/daemon.go]: So, I thought this solution would be fine... but I've changed my mind. We have _way_ too many special-purpose daemon flags that should be settings. We're already trying to get rid of the existing daemon flags, I don't want to add more that we'll need to maintain into the future for backwards compatibility.

If we want to be able to override config options, we need a coordinated system to do that but this isn't that.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:34]: Any reason this can't be `/ip4/127.0.0.1/tcp/0` we can then get the peer's "real" address by looking at the result of `ipfs.Swarm().LocalAddrs(ctx)`

Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go]: Did not know that was an option.  Nice.

Done.

**Comment by gammazero** [comment]
@aschmahmann If you merge PR #8076, then I can rebase and use CAR files to import downloaded migrations.  I can also do that in a separate PR.  In any case, once #8076 is in, import via CAR will be simple.  Making that change to use CAR files will also enable PR #7658 to do the same.

**Comment by aschmahmann** [comment]
@gammazero yep, that's the idea. That PR seems close to being good, but has a bug causing it to fail sharness at the moment. In any event I wouldn't block on that PR here.

File: cmd/ipfs/daemon.go
Comment (gammazero) [reply cmd/ipfs/daemon.go:296]: @Stebalien Done.  Now reading migration setting from config file, and using defaults for any values not present.  Removed CLI flags for specifying migration configuration.

File: cmd/ipfs/migration.go
Comment (aschmahmann) [inline cmd/ipfs/migration.go:61]: Why?

File: cmd/ipfs/migration_test.go
Comment (aschmahmann) [inline cmd/ipfs/migration_test.go:72]:
```suggestion
		t.Fatal("expected 3 fetchers in MultiFetcher")
```

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:239]: What's the idea behind where the background context is getting used and where the passed in context is getting used?

If this function doesn't complete then the temporary node is useless right (aside from the fact that if MDNS is working then everything is probably fine)?

If we don't get the addresses do we need to error + close the node?

File: cmd/ipfs/migration.go
Comment (gammazero) [reply cmd/ipfs/migration.go:61]: Because we state that if they leave the DownloadSources empty, that results in the default behavior.  I thought it might be considered reasonable by some to think `[""]` means empty.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:242]: The background context is used as a mechanism to shutdown the temporary node. Quote from `go-ipfs/core/builder.go`:
>      // Shut down the application if the lifetime context is canceled.
>      // NOTE: we _should_ stop the application by calling `Close()` 
>      // on the process. But we currently manage everything with contexts.                                   

So, the background context is for now shutdown, in the absence of a `node.Close()` method, and the passed-in context is used for cancellation of the current function call.  

True, if the function does not complete, then the node is useless.  However, failure to get the local swarm address only means that the downloaded migrations cannot be fetched from the temp node.... So, right, it does not need to fail out here.

File: cmd/ipfs/migration_test.go
Comment (gammazero) [reply cmd/ipfs/migration_test.go]: Fixed

**Review by aschmahmann — CHANGES_REQUESTED** [review]

File: cmd/ipfs/migration.go
Comment (aschmahmann) [reply cmd/ipfs/migration.go:200]: Given that we just merged CAR export on gateways (https://github.com/ipfs/go-ipfs/pull/8111) I feel much less strongly about this and think we can leave it as it is.

What we can do is in a later RC (or next release) switch the HTTP Fetcher to really be a CAR fetcher and then this problem goes away. For the time being we control how the migrations are created so we can just leave the defaults.

cc @Stebalien as a heads up

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go:79]: Yeah, this is terrible.  Don't know what the plan was here.  This is fixes by never changing the data structure when trying different fetchers.  See next comment.

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go]: I do not change the order, note a preference, or mark any fetchers unusable.  Each fetch just tries them in order. When the daemon tries to add migration downloads locally, it also walks through all the fetchers and gets any downlodads from each.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go]: It does work.  Since the temporary download node is using the same code as the daemon, any plugins that work for one will work for the other. We do not load the plugins twice because the error that results in handled.

Is it necessary?... Probably only for testing.  As long as the IpfsFetcher is used from within a running ipfs daemon, this is not needed as the plugins are already loaded.  If we wanted to use IpfsFetcher outside of the daemon, then we do need to load the plugins, or we get an error like `failed to initialize ephemeral node: unknown datastore type: flatfs`.

Calling `setupPlugins` when plugins are already loaded seems harmless, but should not be necessary if it is only ever called from within the daemon.  Therefore, I moved it out to the test code.

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher_test.go]: Added `skipUnlessEpic` function that works the same way as in addcat_teat.go

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:151]: This was meant to be without a mutex, since it must only be read after all fetching is complete.  I have added the mutex for consistency, but if it were needed here then that is really a misuse of this function.

File: cmd/ipfs/migration.go
Comment (gammazero) [reply cmd/ipfs/migration.go:205]: Let me know if you want me to remove the support for adding downloads from HTTP to IPFS.  ...since the CAR download is on the horizon, and less code to worry about.

File: cmd/ipfs/daemon.go
Comment (gammazero) [reply cmd/ipfs/daemon.go:324]: This is done in a deferred function here, so that cleanup will happen if the daemon returns early due to an error.  It is done later after finishing migration work, so that that directory is not sitting around for the lifetime of the daemon and is not left behind if the daemon exits ungracefully.

Added comment to explain.

File: repo/fsrepo/migrations/ipfsfetcher_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsfetcher_test.go]: This was handled by adding a `skipUnlessEpic` function like as is done here:
 https://github.com/ipfs/go-ipfs/blob/ef866a1400b3b2861e5e8b6cc9edc8633b890a0a/test/integration/addcat_test.go#L173?

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go
Comment (aschmahmann) [reply repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:151]: Makes sense, we may need to revisit this later

**Review by aschmahmann — APPROVED** [review]

**Pull request merged**
